### PR TITLE
Bug 1949840: Improve update and status reporting

### PR DESCRIPTION
**Pull request by jan--f**
This PR adds Three commits, that aim to improve our reconcile loop:

1. Update `prometheus-operator` first and wait for it to finish before updating the other operands concurrently.
2. Allow the reconcile loop three attempts before calling `SetFailed` and thus reporting `available=false`.
3. Back-off more aggressively in case of failures. The `DefaultControllerRateLimiter`s default back-off is a bit slow and it also adds per bucket back-off tracking, which we don't use due to only ever queuing one event. Instead this introduces an instance of `ItemExponentialFailureRateLimiter`, which is also used by `DefaultControllerRateLimiter`, but with more aggressive rate limiting settings.

* [ ] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

File: pkg/operator/operator.go [hunk 1]
```diff
@@ -469,18 +469,30 @@ func (o *Operator) sync(key string) error {
 		klog.Errorf("error occurred while setting status to in progress: %v", err)
 	}
 
+	promOpTaskName := "Updating Prometheus Operator"
```
Comment (simonpasquier) [inline pkg/operator/operator.go:472]: what if `TaskRunner` can take an ordered list of tasks? e.g. 

```golang
tl := NewTaskRunner(client *client.Client)

# first group of tasks to execute.
tl.AddTasks(
    tasks.NewTaskSpec("Updating Prometheus Operator", tasks.NewPrometheusOperatorTask(o.client, factory)),
)
# second group of tasks to execute.
tl.AddTasks(
    tasks.NewTaskSpec("Updating user workload Prometheus Operator", tasks.NewPrometheusOperatorUserWorkloadTask(o.client, factory, config)),
    ...
)

taskName, err := tl.RunAll()
```

File: pkg/operator/operator.go [hunk 2]
```diff
@@ -489,6 +501,16 @@ func (o *Operator) sync(key string) error {
 	return nil
 }
 
+func (o *Operator) reportError(err error, taskName string) error {
+	if o.failedReconcileAttempts < 3 {
+		o.failedReconcileAttempts++
+		return nil
+	}
```
Comment (simonpasquier) [inline pkg/operator/operator.go:508]: should we reset `o.failedReconcileAttempts`?

Comment (jan--f) [reply pkg/operator/operator.go:508]: No, I think we only want to do reset once we were successful. This happens [here](https://github.com/openshift/cluster-monitoring-operator/pull/1193/files#diff-99c0290799daf9abc6240df64063e20bfaf67b371577b67ac7eec6f4725622ffR495)

File: pkg/operator/operator.go (hunk 1 above)
Comment (jan--f) [reply pkg/operator/operator.go:472]: Thought about it, but figured as long as we only have one task to run before, I might as well use whats already there.

Comment (simonpasquier) [reply pkg/operator/operator.go:469]: Thinking about it more, I'd rather see us only reuse TaskRunner (even for only one task):

```golang
# first group of tasks to execute
tl := NewTaskRunner(o.client,[]*tasks.TaskSpec{
    tasks.NewTaskSpec("Updating Prometheus Operator", tasks.NewPrometheusOperatorTask(o.client, factory)),
})

taskName, err := tl.RunAll()
...

# second group of tasks to execute.
tl = NewTaskRunner(o.client,[]*tasks.TaskSpec{
    tasks.NewTaskSpec("Updating user workload Prometheus Operator", tasks.NewPrometheusOperatorUserWorkloadTask(o.client, factory, config)),
    ...
)

taskName, err = tl.RunAll()
...
```

File: pkg/operator/operator.go (hunk 2 above)
Comment (simonpasquier) [reply pkg/operator/operator.go:503]: So the operator would retry up to 3 times before reporting available=false/degraded=true and the next sync() operations wouldn't retry?

```
Updating ClusterOperator status to failed after retrying 3 times. Err: xxx
...
Updating ClusterOperator status to failed after retrying 4 times. Err: xxx
...
Updating ClusterOperator status to failed after retrying 5 times. Err: xxx
```

File: pkg/operator/operator.go (hunk 1 above)
Comment (jan--f) [reply pkg/operator/operator.go:469]: Opted for the first idea of a list of task lists. This way we don't need to add error handling if the first task list has a failed task.

File: pkg/operator/operator.go (hunk 2 above)
Comment (jan--f) [reply pkg/operator/operator.go:503]: The intention is that it allows for 3 failed attempts before reporting `available=false/degraded=true`. All error cause a retry to sync the event in question though. `o.failedReconcileAttempts` is really only affecting if/when the operator status changes...unless I'm missing something.

The status and `o.failedReconcileAttempts` is reset to `0` and `available=true/degraded=false` respectively once a call to sync completes without error

File: pkg/operator/operator.go [hunk 3]
```diff
@@ -170,7 +172,7 @@ func New(
 		namespace:                 namespace,
 		namespaceUserWorkload:     namespaceUserWorkload,
 		client:                    c,
-		queue:                     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "cluster-monitoring"),
+		queue:                     workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(50*time.Millisecond, 3*time.Minute), "cluster-monitoring"),
```
Comment (ianbillett) [inline pkg/operator/operator.go:175]: this change doesn't appear to be related to the stated goals of the PR? This reads like a different back-off mechanism?
> * Update prometheus-operator first and wait for it to finish before updating the other operands concurrently.
> * Allow the reconcile loop three attempts before calling SetFailed and thus reporting available=false.

File: pkg/tasks/tasks.go [hunk 4]
```diff
@@ -21,43 +21,53 @@ import (
 	"k8s.io/klog/v2"
 )
 
+// TaskRunner executes groups of tasks, the groups sequentially, each group concurrently
```
Comment (ianbillett) [inline pkg/tasks/tasks.go:24]: nit: semantically, the struct itself does not execute the group of tasks -I would guess the `RunAll` function does that?

**Review by ianbillett — COMMENTED** [review]
I don't have much context so a couple of points about readability :)

File: pkg/operator/operator.go (hunk 3 above)
Comment (jan--f) [reply pkg/operator/operator.go:175]: Ah yes, I'll update the PR description. The respective [commit](https://github.com/openshift/cluster-monitoring-operator/pull/1193/commits/62e2d9dcd878389fb7d8e368e2b8727bf845da9f) has details.

Thanks, good catch.

File: pkg/tasks/tasks.go (hunk 4 above)
Comment (jan--f) [reply pkg/tasks/tasks.go]: Fair point, add more comments and shuffled stuff around.

File: pkg/tasks/tasks.go [hunk 5]
```diff
@@ -21,43 +21,60 @@ import (
 	"k8s.io/klog/v2"
 )
 
+// TaskRunner manages lists of tasks. Through the RunAll method tasks are
+// executed, the lists sequentially, each list of tasks concurrently.
 type TaskRunner struct {
-	client *client.Client
-	tasks  []*TaskSpec
+	client    *client.Client
+	taskLists [][]*TaskSpec
```
Comment (fpetkovski) [inline pkg/tasks/tasks.go:28]: What is the reason for changing the type here?

File: pkg/tasks/tasks.go [hunk 6]
```diff
@@ -21,43 +21,60 @@ import (
...
 }
 
+// NewTaskRunner returns a task runner. tasks is the first task list that will
```
Comment (fpetkovski) [inline pkg/tasks/tasks.go:31]: What do you mean by `tasks is the first task`?

Comment (jan--f) [reply pkg/tasks/tasks.go:31]: `tasks` is referring to the function argument. And it is the first task list that gets processed.

File: pkg/tasks/tasks.go (hunk 5 above)
Comment (jan--f) [reply pkg/tasks/tasks.go:28]: Does the [commit message for this change](https://github.com/openshift/cluster-monitoring-operator/pull/1193/commits/d6d1510c5253e1ec2cdeedf3b2d5bbbb8345697f) make things clearer?

Part of this PR is to extend `TaskRunner` to process multiple groups of tasks as a way to enforce some order.

Comment (fpetkovski) [reply pkg/tasks/tasks.go:28]: Ah yes, that makes sense. To make sure this is reflected in the codebase, can we introduce a new abstraction (e.g. `TaskGroup`) where we can document this as part of the godoc?

Comment (fpetkovski) [reply pkg/tasks/tasks.go:28]: This is definitely just optional and if it is too big of a change it's fine to keep the existing solution.

Comment (simonpasquier) [reply pkg/tasks/tasks.go:28]: I agree with @fpetkovski that it might be good to introduce another struct like `TaskGroup` for clarity.

File: pkg/operator/operator.go (hunk 2 above)
Comment (simonpasquier) [reply pkg/operator/operator.go:500]: :+1:

File: pkg/operator/operator.go [hunk 7]
```diff
@@ -489,6 +489,18 @@ func (o *Operator) sync(key string) error {
 	return nil
 }
 
+func (o *Operator) reportError(err error, taskName string) {
+	if o.failedReconcileAttempts > 3 {
+		klog.Infof("Updating ClusterOperator status to failed after retrying %v times. Err: %v", o.failedReconcileAttempts, err)
```
Comment (simonpasquier) [inline pkg/operator/operator.go:494]: I might be worth logging the error message at every step? IIUC the first 3 errors would be missing currently.

File: pkg/tasks/tasks.go [hunk 8]
```diff
@@ -21,35 +21,64 @@ import (
...
-	client *client.Client
-	tasks  []*TaskSpec
+	client     *client.Client
+	taskGroups []*TaskGroup
 }
 
-func NewTaskRunner(client *client.Client, tasks []*TaskSpec) *TaskRunner {
+// NewTaskRunner returns a task runner. tasks is the first task group that will
+// be executed, before any list added via AppendTaskGroup.
+func NewTaskRunner(client *client.Client, tasks *TaskGroup) *TaskRunner {
```
Comment (simonpasquier) [inline pkg/tasks/tasks.go:33]: We could make it a variadic function and get rid of `AppendTaskGroup`?

```suggestion
func NewTaskRunner(client *client.Client, taskGroups ...*TaskGroup) *TaskRunner {
```

File: pkg/operator/operator.go [hunk 9]
```diff
@@ -489,6 +491,20 @@ func (o *Operator) sync(key string) error {
 	return nil
 }
 
+func (o *Operator) reportError(err error, taskName string) {
```
Comment (simonpasquier) [inline pkg/operator/operator.go:494]: (nit) small suggestion to make it easier to read IMHO

```golang
func (o *Operator) reportError(err error, taskName string) {
    defer o.failedReconcileAttempts++

    klog.Infof("Operator reconciliation failed, retrying. Err: %v", err)
    if o.failedReconcileAttempts < 3 {
        // Don't update the ClusterOperator status until 3 retries have been attempted.
        return
    }

    ...
```

File: pkg/tasks/tasks.go [hunk 10]
```diff
@@ -21,35 +21,64 @@ import (
...
 		// shadow vars due to concurrency
 		ts := ts
 		i := i
 
 		g.Go(func() error {
-			klog.V(2).Infof("running task %d of %d: %v", i+1, len(tl.tasks), ts.Name)
-			err := tl.ExecuteTask(ts)
-			klog.V(2).Infof("ran task %d of %d: %v", i+1, len(tl.tasks), ts.Name)
+			klog.V(2).Infof("running task %d of %d: %v", i+1, tgLength, ts.Name)
+			err := ts.Task.Run()
```
Comment (simonpasquier) [inline pkg/tasks/tasks.go:73]: I suppose that `ExecuteTask()` can be removed then?

File: pkg/tasks/tasks.go (hunk 8 above)
Comment (jan--f) [reply pkg/tasks/tasks.go:33]: :+1: Indeed we could...and probably should too.

File: pkg/tasks/tasks.go (hunk 10 above)
Comment (jan--f) [reply pkg/tasks/tasks.go:73]: Good catch, yes that was meant to be removed.

File: pkg/operator/operator.go (hunk 9 above)
Comment (jan--f) [reply pkg/operator/operator.go:494]: Hm only that `defer` doesn't work. It expects a function call, whereas we have a statement. Not sure wrapping it in a function helps readability here.

File: pkg/operator/operator.go [hunk 11]
```diff
@@ -441,26 +439,31 @@ func (o *Operator) sync(key string) error {
 	}
 	factory := manifests.NewFactory(o.namespace, o.namespaceUserWorkload, config, o.loadInfrastructureConfig(), proxyConfig, o.assets)
 
+	// update prometheus-operator before anything else
```
Comment (simonpasquier) [inline pkg/operator/operator.go:442]: Maybe move this right before the first `tasks.NewTaskGroup`?

```suggestion
	// Update prometheus-operator before anything else because it is responsible for managing many other resources (e.g. Prometheus, Alertmanager, Thanos Ruler, ...).
```

File: pkg/operator/operator.go (hunk 9 above)
Comment (simonpasquier) [reply pkg/operator/operator.go:491]: my bad I didn't verify that my pseudo code worked :) maybe we could still split the logs (1 for logging the error and the other telling that we update the cluster operator status)?

```golang
    klog.Infof("Reconciliation failed (%d retries). Err: %v", o.failedReconcileAttempts, err)
    if o.failedReconcileAttempts > 3 {
        // Only update the ClusterOperator status after 3 retries have been attempted to avoid flapping status.
        klog.Info("Updating ClusterOperator status to failed ")
        ...
    }
```

Comment (jan--f) [reply pkg/operator/operator.go:491]: Changes the logging as per your comment. I also changed the condition to 2 (we start counting at 0 so `> 3` was incorrect and to an equality check. No need to set `degraded` repeatedly I figure.
